[tests] workload tests for test project templates

radical opened the pull request:
- Add tests for test framework specific templates - `aspire-xunit`, `aspire-nunit`, and `aspire-mstest`.
- Also, test these with different project names to validate the generated code
- And add a test for validating a `aspire-starter` template with the various project names
- The tests revealed a missing `using` in the test cs file, fixed here

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4240)

radical committed:
Fix xunit template build

```
... testroot/aspire-aspire-xunit/aspire-aspire-xunit.aspire-xunitTests/IntegrationTest1.cs(21,27): error CS1061:
'IServiceCollection' does not contain a definition for 'ConfigureHttpClientDefaults' and no accessible extension method 'ConfigureHttpClientDefaults' accepting a first argument of type 'IServiceCollection' could be found (are you missing a using directive or an assembly reference?) [/datadisks/disk1/work/B74F0975/t/testroot/aspire-aspire-xunit/aspire-aspire-xunit.aspire-xunitTests/aspire-aspire-xunit.aspire-xunitTests.csproj]
```

radical commented:
@DamianEdwards the xunit test file in the template seems to be broken. I think this should fix it.

[`edd5812` (#4240)](https://github.com/dotnet/aspire/pull/4240/commits/edd581268b64a1868d72ddcc0b673ccf55e63c30)

The test failures: https://dev.azure.com/dnceng-public/public/_build/results?buildId=708682&view=ms.vss-test-web.build-test-results-tab

radical committed:
[tests] Workload: add tests for aspire-{testFramework} templates

radical committed:
[tests] Workload: Add check to ensure that all the expected nugets are produced. Useful for local builds

radical committed:
fix typo

radical committed:
disable failing tests, and reference the relevant issues

radical committed:
fix

radical committed:
Merge remote-tracking branch 'origin/main' into workload-tests-follow-up

joperezr commented on tests/Shared/Aspire.Workload.Testing.props:
```diff
@@ -2,6 +2,7 @@
   <Import Project="$(RepoRoot)src\Shared\Workload.props" />
 
   <PropertyGroup>
+    <InstallWorkloadUsingArtifactsAfterThisTarget />
```
NIT: Can we add a comment here explaining why this property needs to get reset.

joperezr commented on tests/Shared/Aspire.Workload.Testing.targets:
```diff
@@ -169,4 +169,68 @@
+            }
+
+            var files = Directory.EnumerateFiles(SourceDirectory, "*.csproj", SearchOption.AllDirectories);
+            PackageNames = files.Where(f => File.ReadAllText(f).Contains("<IsPackable>true</IsPackable>"))
```
This would break if we move the IsPackable to a Directory.Build.props file, right? I wonder if instead of using an inline task if it's best to just use MSBuild task for all of these projects and call a target that just returns the value of IsPackable. We don't have to do it that way but I just think that would probably be more future proof and resilient against refactorings.

radical replied:
Thanks for the suggestion! Changed.

joperezr reviewed (approved):
Left couple of suggestions, but looks good otherwise.

radical committed:
address review feedback from @ joperezr

radical committed:
Merge remote-tracking branch 'origin/main' into workload-tests-follow-up

eerhardt commented on tests/Aspire.Workload.Tests/PerTestFrameworkTemplatesTests.cs:
```diff
@@ -0,0 +1,179 @@
+
+                if (inTest && CommentLineRegex().IsMatch(line))
+                {
+                    sb.AppendLine(CommentLineRegex().Replace(line, "    "));
```
Hopefully we don't ever add real comments to the code, or else this would start failing.

radical replied:
Lol! Yeah, but that would cause a failure on the PR, and then that person gets to figure it out :/

eerhardt commented on tests/Aspire.Workload.Tests/PerTestFrameworkTemplatesTests.cs:
```diff
@@ -0,0 +1,179 @@
+        _testTemplateName = testType;
+    }
+
+    public static TheoryData<string> ProjectNamesWithTestTemplate_TestData()
+    {
+        var data = new TheoryData<string>();
+        foreach (var name in GetProjectNamesForTest())
+        {
+            data.Add(name);
+        }
+        return data;
+    }
```
This could be simplified:

```suggestion
    public static IEnumerable<object[]> ProjectNamesWithTestTemplate_TestData()
    {
        foreach (var name in GetProjectNamesForTest())
        {
            yield return [name];
        }
    }
```

radical committed:
Update tests/Aspire.Workload.Tests/WorkloadTestsBase.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
Update tests/Aspire.Workload.Tests/PerTestFrameworkTemplatesTests.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
WorkloadTesting.targets: Use MSBuild task to check whether projects are

.. packable, instead of a custom task. Doing this instead of a custom
task looking for a `<IsPackable>true</IsPackable>` string is better
because it would not be affected by refactorings, or the property being set
elsewhere.

Suggestion from @jopererz .

radical committed:
Merge remote-tracking branch 'origin/main' into workload-tests-follow-up

DamianEdwards commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -180,6 +182,53 @@ protected static async Task<ResourceRow[]> CheckDashboardHasResourcesAsync(IPage
+    {
+        if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+        {
+            // ActiveIssue for windows: https://github.com/dotnet/aspire/issues/4555
+            yield return "aspire_龦唉丂荳_㐁ᠭ_ᠤསྲིདخەلꌠ_1ᥕ";
+        }
+
+        // ActiveIssue: https://github.com/dotnet/aspire/issues/4550
+        // yield return "aspire  sta-rter.test"; // Issue: two spaces
+
+        yield return "aspire_starter.1period then.34letters";
+        yield return "aspire-starter & with.1";
+
+        // ActiveIssue: https://github.com/dotnet/aspnetcore/issues/56277
+        // yield return "aspire_😀";
```
I think falls under "naming is one of the hardest problems in computer science" right? 😉

radical replied:
For the curious this might be a bug in the template engine - https://github.com/dotnet/sdk/issues/41667 !
